File: 0744-network-delay-time/0744-network-delay-time.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;
        //{min_dist,min_node}
        vector<vector<pair<int,int>>> adjList(n+1);
        for(auto time:times) {
            adjList[time[0]].push_back({time[1],time[2]});
        }


        vector<int> dist(n+1,1e5);
        dist[k]=0;
        pq.push({0,k});
        while(!pq.empty()) {
            pair<int,int> top = pq.top();
            pq.pop();
            for(auto node:adjList[top.second]) {
                int neighbour = node.first;
                int wt = node.second;
                if(dist[top.second]+wt<dist[neighbour]) {
                    dist[neighbour] = dist[top.second]+wt;
                    pq.push({dist[neighbour],neighbour});
                }
            }
        }
        int maxDist = *max_element(dist.begin()+1,dist.end());
        
        if (maxDist==1e5) return -1;
        return maxDist;
        

    }
};

File: 0744-network-delay-time/0744-network-delay-time.py
import heapq
from collections import defaultdict

class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # Initialize distances to infinity for all nodes
        dist = [float("inf")] * (n + 1)
        dist[k] = 0  # Distance to the starting node is 0
        
        # Priority queue to process nodes with the shortest distance
        pq = [(0, k)]  # (distance, node)
        
        # Create adjacency list for the graph
        adjList = defaultdict(list)
        for time in times:
            u, v, w = time
            adjList[u].append((v, w))
        
        # Dijkstra's algorithm
        while pq:
            curDist, node = heapq.heappop(pq)
            
            # Skip processing if the current distance is not optimal
            if curDist > dist[node]:
                continue
            
            # Explore neighbors
            for neighbor, neighborCost in adjList[node]:
                if dist[node] + neighborCost < dist[neighbor]:
                    dist[neighbor] = dist[node] + neighborCost
                    heapq.heappush(pq, (dist[neighbor], neighbor))
        
        # Calculate the maximum time to reach all nodes
        maxDist = max(dist[1:])  # Ignore index 0 as nodes are 1-indexed
        return maxDist if maxDist != float("inf") else -1
